server/src/main/java/org/opensearch/wlm/cancellation/DefaultTaskCancellation.java

```diff
@@ -0,0 +1,277 @@
+        this.queryGroupLevelResourceUsageViews = queryGroupLevelResourceUsageViews;
+        this.activeQueryGroups = activeQueryGroups;
+        this.deletedQueryGroups = deletedQueryGroups;
+        this.isNodeInDuress = isNodeInDuress;
+    }
+
+    /**
+     * Cancel tasks based on the implemented strategy.
+     */
+    public final void cancelTasks() {
+        // cancel tasks from QueryGroups that are in Enforced mode that are breaching their resource limits
+        cancelTasks(QueryGroup.ResiliencyMode.ENFORCED);
+        // if the node is in duress, cancel tasks accordingly.
+        handleNodeDuress();
+    }
+
+    private void handleNodeDuress() {
+        if (!isNodeInDuress.getAsBoolean()) {
+            return;
+        }
+        // List of tasks to be executed in order if the node is in duress
+        List<Consumer<Void>> duressActions = List.of(
+            v -> cancelTasksFromDeletedQueryGroups(),
+            v -> cancelTasks(QueryGroup.ResiliencyMode.SOFT)
+        );
+
+        for (Consumer<Void> duressAction : duressActions) {
+            if (!isNodeInDuress.getAsBoolean()) {
+                break;
+            }
+            duressAction.accept(null);
+        }
+    }
+
+    private void cancelTasksFromDeletedQueryGroups() {
+        cancelTasks(getAllCancellableTasks(this.deletedQueryGroups));
+    }
+
+    /**
+     * Get all cancellable tasks from the QueryGroups.
+     *
+     * @return List of tasks that can be cancelled
+     */
+    protected List<TaskCancellation> getAllCancellableTasks(QueryGroup.ResiliencyMode resiliencyMode) {
+        return getAllCancellableTasks(getQueryGroupsToCancelFrom(resiliencyMode));
+    }
+
+    /**
+     * Get all cancellable tasks from the given QueryGroups.
+     *
+     * @return List of tasks that can be cancelled
+     */
+    protected List<TaskCancellation> getAllCancellableTasks(Collection<QueryGroup> queryGroups) {
+        return queryGroups.stream().flatMap(queryGroup -> getCancellableTasksFrom(queryGroup).stream()).collect(Collectors.toList());
+    }
+
+    /**
+     * returns the list of QueryGroups breaching their resource limits.
+     *
+     * @return List of QueryGroups
+     */
+    private List<QueryGroup> getQueryGroupsToCancelFrom(QueryGroup.ResiliencyMode resiliencyMode) {
+        final List<QueryGroup> queryGroupsToCancelFrom = new ArrayList<>();
+
+        for (QueryGroup queryGroup : this.activeQueryGroups) {
+            if (queryGroup.getResiliencyMode() != resiliencyMode) {
+                continue;
+            }
+            Map<ResourceType, Long> queryGroupResourceUsage = queryGroupLevelResourceUsageViews.get(queryGroup.get_id())
+                .getResourceUsageData();
+
+            for (ResourceType resourceType : TRACKED_RESOURCES) {
+                if (queryGroup.getResourceLimits().containsKey(resourceType) && queryGroupResourceUsage.containsKey(resourceType)) {
```

Comment:
I think right side of this && operator is not required

Reply:
its just a check to proceed with `isBreachingThreshold` only if the querygroup has limits set on resourceType and there is usage recorded for the usage type.

Reply:
That is true but it is unnecessary because the `queryGroupResourceUsage` map contains resource usage for all tracked resources